### ISSUE TITLE
Dynamic imports with System.import()

### ISSUE THREAD

**Issue by mover96** [rollup/rollup#787]
Hi this is more of a question/clarification than an issue.

Coming from webpack, I find it handles statements like `System.import('./foo.js');` without issue (and I am sure Rollup could too). However in my project I have a `System.import` that uses objects that the server attaches to `Window` to dynamically import Angular components. This is where webpack falls apart, as it cannot know what it will be importing until runtime, so it does not know what to bundle (it does have some fancy context stuff where it tries to guess, but since my paths vary greatly it tries to bundle practically everything and it fails). I would imagine this is an issue that every bundler will have, as it is impossible to know what to bundle until runtime. 

But I thought I would ask anyways, can Rollup handle something like this? Unfortunately redesigning my app is not an option currently, but I am open to suggestions regardless as it would be good to know for the future. Since ES6 supports these dynamic imports, I can see this becoming a larger issue as bundlers gain more traction (although I admit that importing modules via `Window` isn't the most elegant approach).

**Comment by Victorystick** [comment]
I'm not quite sure I follow you. Do you know what objects the server has attached to the window? If you do, why not just access them right-off-the-bat in the bundle; `window.serverArrachedObject`?

**Comment by mover96** [comment]
Sorry I don't think I explained myself fully! The objects that are on the window contain URLs that point to angular components (i.e. *.ts files) that are then requested to be loaded. One of the issues is that they are scattered throughout the app, the highest level folder they all share is the app itself, and since my app is fairly large I can't just bundle everything. 

The way this worked before bundling was we had tsc convert every single ts file to js and exported them to the disk, and then used systemjs to import these files dynamically (as the URL for the file is only available at runtime since they are attached to the window by the server). 

But with bundling, even if I write an import statement for all 60 files I want to import (which I would also like to try and avoid), as far as I understand it bundling kinda overwrite the relative pathing. So when system.import gets the url from the window object at runtime, i.e. "path/to/file/A", even though files A is imported, it doesn't understand where that path points in the bundle. And even if this way is possible I need to remove the need for a static import for file A because I want them to be imported dynamically.

So that goes back to the original issue of me not being able to bundle the files without knowing their paths which sounds impossible with bundling. It is just kind of awkward because ES6 brings functionalities like this, yet it practically breaks bundlers. Instead of having to alter my source files for bundlers, I would like to be able to alter a config file and perhaps point rollup towards every possible dependency (which is still kind of frustrating because right now we use server side discovery to find the files so no one has to add a line to a file each time a new one is created, but I digress) so that rollup could be aware of the files, and when the URLs are resolved at runtime it is expecting those URLs and points them to the appropriate bundled file/function or whatever.

Does this make anymore sense? I am quite new to all of this so this is how my current understanding of everything is!

**Comment by Victorystick** [comment]
What do you do with the loaded Angular components? Are you only importing importing the files for side-effects? Do the components register themselves when loaded? It wouldn't make sense to import them, without doing anything with them. I'm sorry. I don't follow how the flow of the application.

**Comment by mover96** [comment]
No, I import the components and then basically "bootstrap" them with Angular's dynamicComponentLoader. So they are used fully since they are independent components.

**Comment by mover96** [comment]
The angular aspect isn't so important, I just need to import a module, and use it in the application, without knowing what the path will be until runtime. It works with native javascript, but the bundler is what cannot handle that kind of dependency.

**Comment by mover96** [comment]
Here is what the import statement looks like so you can get a better idea of what is going on:
```
System.import(path)
            .then(m => {
                this._dynamicComponentLoader.loadNextToLocation(m[this.sampleName], this._viewContainerRef)
                    .then(component => {
                        this._changeDetectorRef.markForCheck();
                        if (this.componentDetails.type === 'global') {
                            this._http.get(`/styleguide/ct/${this.componentDetails.name}/${sampleFileName}`)
                                .subscribe(html => {
                                    this.markup = this._styleguideMetadataService.stripStyleguideMetadata(html.text());
                                    this._changeDetectorRef.markForCheck();
                                });
                        }
                    });
            });
```
`path` is what is taken from the window object, along with `sampleName`

**Comment by mover96** [comment]
Just in case anyone stumbles upon this, this is possible to do with webpack. See https://github.com/webpack/webpack/issues/2783

**Comment by bfred-it** [comment]
Closing this because:

1. I haven't seen `System.import()` a lot in the wild and no one else asked for it in the past year
2. > Implementing System.import is, however, out-of-scope for Rollup.
  > @Victorystick (one of rollup owners) in https://github.com/rollup/rollup/issues/557#issuecomment-210096228
3. `import()` is more likely to be implemented as part of #372
